src/Core/Checkout/Cart/Order/RecalculationService.php

```diff
@@ -360,9 +360,10 @@ private function recalculateCart(Cart $cart, SalesChannelContext $context): Cart
             $cart = $this->processor->process($cart, $live, $behavior);
 
             // validate cart against the context rules
-            $validated = $this->cartRuleLoader->loadByCart($live, $cart, $behavior);
+            $validatedCart = $this->cartRuleLoader->loadByCart($live, $cart, $behavior)->getCart();
+            $validatedCart->addErrors(...$cart->getErrors());
```

Comment:
sorry for not mentioning this earlier, but persistent errors will be taken over anyway, which could lead to duplication 😅:
```suggestion
            $validatedCart->addErrors(...$cart->getErrors()->filter(fn ($error) => !$error->isPersistent()));
```